numba/_numba_common.h

```diff
@@ -17,4 +17,21 @@
 #define VISIBILITY_HIDDEN
 #endif
 
+/*
+ * Numba's version of the PyArray_DescrCheck macro from NumPy, use it as a
+ * direct replacement of NumPy's PyArray_DescrCheck to ensure binary
+ * compatibility.
+ *
+ * Details of why this is needed:
+ * NumPy 1.18 changed the definition of the PyArray_DescrCheck macro here:
+ * https://github.com/numpy/numpy/commit/6108b5d1e138d07e3c9f2a4e3b1933749ad0e698
+ * the result of this being that building against NumPy <1.18 would prevent
+ * Numba running against NumPy >= 1.20 as noted here:
+ * https://github.com/numba/numba/issues/6041#issuecomment-665132199
+ *
+ * This macro definition is copied from:
+ * https://github.com/numpy/numpy/commit/6108b5d1e138d07e3c9f2a4e3b1933749ad0e698#diff-ad2213da23136c5fc5883d9eb2d88666R26
+ */
+#define NUMBA_PyArray_DescrCheck(op) PyObject_TypeCheck(op, &PyArrayDescr_Type)
```

Comment:
Possibly worth emphasizing that this is the _new_ definition of the macro. Perhaps better would be to do something like
```
#if numpy_is_too_old
    #undef PyArray_DescrCheck
    #define PyArray_DescrCheck ...
#endif
```
@seberg, is there an easy way to do the `numpy_is_too_old` check?

Reply:
Hmmmpf, we do have `C_API_VERSION`, but I am not sure it incremented in 1.18 specifically. Is the actual numpy version available in the header? I can't find it quickly, can't say I ever looked into those things.

The new definition is valid also for old numpy versions though.

Reply:
hmmm, there's `NPY_<X>_<Y>_API_VERSION`, `NPY_ABI_VERSION` and `NPY_API_VERSION`, but if they haven't been moved I'm not sure it's something to pursue?

Perhaps the best thing is to update the comment as @eric-wieser suggested and say that this is the new definition and it's backwards compatible?

Reply:
The `NPY_<X>_<Y>_API_VERSION` macro seems strange, we do have a C-API version, independend of the numpy version, which may even have incremented in 1.18.x, but that macro seems to have never changed, so I am unsure what it actually represents.

Reply:
@seberg any idea where the C-API version is/what it's called?

Reply:
looks like the old definition is https://github.com/numpy/numpy/blob/maintenance/1.15.x/numpy/core/include/numpy/ndarrayobject.h#L32
the new code would be compatible with the old definition. I don't see the need to have it conditional on API_VERSION, which we ain't sure.

Reply:
> I don't see the need to have it conditional on API_VERSION, which we ain't sure.

My suggestion was more a case of guarding against it changing again - but maybe that's unlikely and/or not relevant.
The approach in the patch is fine too, with a suitable comment.

Reply:
I think the current definition will always stay correct.  We might be able to define some faster method/path, but I doubt that is relevant or even likely.  And even if we define a faster way, we would have to wait until nobody uses pre 1.20...

Quite honestly, I think you could also just not use a macro and write out the TypeCheck...

Reply:
Note about it being 1.18 and above version of the macro added in https://github.com/numba/numba/pull/6078/commits/2d49e4d878f1eab44eceaa7accc3134843c5188e

Reply:
Thanks @eric-wieser @seberg, given the thoughts above, and with the note made in https://github.com/numba/numba/pull/6078/commits/2d49e4d878f1eab44eceaa7accc3134843c5188e, I'm inclined to leave this as-is until something proves it's not sufficient, @sklam what do you think?